File: extension/src/providers/cloud.ts
/**
 * CloudProvider for the browser extension.
 * Supports OpenAI and Gemini APIs with retry logic.
 */
import type { Edit, LlmResponse } from '../types/edit';
import type { Provider } from './index';
import { GRAMMAR_CHECK_SYSTEM_PROMPT, formatUserMessage } from './prompts';

const JSON_SCHEMA = {
  type: 'object',
  required: ['edits'],
  properties: {
    edits: {
      type: 'array',
      items: {
        type: 'object',
        required: ['start_index', 'end_index', 'replacement', 'category'],
        properties: {
          start_index: { type: 'integer', minimum: 0 },
          end_index: { type: 'integer', minimum: 1 },
          replacement: { type: 'string' },
          category: { type: 'string', enum: ['grammar', 'style'] },
          explanation: { type: 'string' },
        },
        additionalProperties: false,
      },
    },
  },
  additionalProperties: false,
};

export class CloudProvider implements Provider {
  readonly providerName: string;
  readonly modelName: string;
  private readonly apiKey: string;
  private readonly type: 'openai' | 'gemini';

  constructor(type: 'openai' | 'gemini', apiKey: string, model?: string) {
    this.type = type;
    this.apiKey = apiKey;
    this.providerName = type;
    this.modelName = model ?? (type === 'openai' ? 'gpt-4o-mini' : 'gemini-2.0-flash');
  }

  async check(text: string): Promise<Edit[]> {
    return this.withRetry(() =>
      this.type === 'openai' ? this.callOpenAI(text) : this.callGemini(text)
    );
  }

  private async withRetry<T>(fn: () => Promise<T>): Promise<T> {
    const maxAttempts = 3;
    let lastError: unknown;
    for (let attempt = 0; attempt < maxAttempts; attempt++) {
      if (attempt > 0) {
        await new Promise((r) => setTimeout(r, (1 << (attempt - 1)) * 1000 + 50));
      }
      try {
        return await fn();
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        const retryable = msg.includes('429') || msg.includes('502') || msg.includes('503');
        if (!retryable) throw err;
        lastError = err;
      }
    }
    throw lastError ?? new Error('All retry attempts failed');
  }

  private async callOpenAI(text: string): Promise<Edit[]> {
    const response = await fetch('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        model: this.modelName,
        messages: [
          { role: 'system', content: GRAMMAR_CHECK_SYSTEM_PROMPT },
          { role: 'user', content: formatUserMessage(text) },
        ],
        response_format: {
          type: 'json_schema',
          json_schema: { name: 'grammar_check', strict: true, schema: JSON_SCHEMA },
        },
      }),
    });

    if (response.status === 401) {
      throw new Error('OpenAI authentication failed. Check your API key in extension settings.');
    }
    if (!response.ok) {
      throw new Error(`OpenAI returned HTTP ${response.status}`);
    }

    const data = (await response.json()) as { choices?: Array<{ message?: { content?: string } }> };
    const content = data.choices?.[0]?.message?.content;
    if (!content) throw new Error('Unexpected OpenAI response shape');

    return parseEdits(content, text);
  }

  private async callGemini(text: string): Promise<Edit[]> {
    const url = `https://generativelanguage.googleapis.com/v1beta/models/${this.modelName}/generateContent?key=${this.apiKey}`;
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        contents: [
          {
            parts: [
              { text: GRAMMAR_CHECK_SYSTEM_PROMPT },
              { text: formatUserMessage(text) },
            ],
          },
        ],
        generationConfig: {
          responseMimeType: 'application/json',
          responseSchema: {
            type: 'OBJECT',
            required: ['edits'],
            properties: {
              edits: {
                type: 'ARRAY',
                items: {
                  type: 'OBJECT',
                  required: ['start_index', 'end_index', 'replacement', 'category'],
                  properties: {
                    start_index: { type: 'INTEGER' },
                    end_index: { type: 'INTEGER' },
                    replacement: { type: 'STRING' },
                    category: { type: 'STRING', enum: ['grammar', 'style'] },
                    explanation: { type: 'STRING' },
                  },
                },
              },
            },
          },
          temperature: 0.1,
        },
      }),
    });

    if (!response.ok) {
      throw new Error(`Gemini returned HTTP ${response.status}`);
    }

    const data = (await response.json()) as {
      candidates?: Array<{ content?: { parts?: Array<{ text?: string }> } }>;
    };
    const content = data.candidates?.[0]?.content?.parts?.[0]?.text;
    if (!content) throw new Error('Unexpected Gemini response shape');

    return parseEdits(content, text);
  }
}

function parseEdits(json: string, originalText: string): Edit[] {
  let parsed: LlmResponse;
  try {
    parsed = JSON.parse(json) as LlmResponse;
  } catch {
    throw new Error(`Invalid JSON from LLM: ${json.slice(0, 200)}`);
  }

  const textLen = originalText.length;
  return parsed.edits.filter((edit) => {
    if (edit.start_index >= edit.end_index) return false;
    if (edit.end_index > textLen) return false;
    if (originalText.slice(edit.start_index, edit.end_index) === edit.replacement) return false;
    return true;
  });
}

File: cli/src/providers/cloud.rs
//! Cloud provider implementations for OpenAI and Gemini.
//!
//! Both providers implement exponential backoff retry logic (3 attempts, 1s/2s/4s).

use anyhow::{Context, Result};
use serde_json::json;
use tokio::time::{sleep, Duration};

use crate::models::config::CloudProvider;
use crate::models::edit::Edit;
use crate::providers::prompts::{format_user_message, GRAMMAR_CHECK_SYSTEM_PROMPT};
use crate::services::validator;

/// Provider that calls cloud LLM APIs (OpenAI or Gemini).
pub struct CloudLlmProvider {
    provider: CloudProvider,
    model: String,
    api_key: String,
    client: reqwest::Client,
}

impl CloudLlmProvider {
    pub fn new(provider: CloudProvider, model: String, api_key: String) -> Self {
        Self {
            provider,
            model,
            api_key,
            client: reqwest::Client::new(),
        }
    }

    /// Executes the API call with exponential backoff retry (3 attempts).
    async fn call_with_retry(&self, text: &str) -> Result<String> {
        const MAX_ATTEMPTS: u32 = 3;
        let mut last_error = None;

        for attempt in 0..MAX_ATTEMPTS {
            if attempt > 0 {
                // Exponential backoff: 1s, 2s, 4s (+ small jitter via offset)
                let wait_ms = (1u64 << (attempt - 1)) * 1000 + 50;
                sleep(Duration::from_millis(wait_ms)).await;
            }

            match self.call_api(text).await {
                Ok(result) => return Ok(result),
                Err(e) => {
                    // Check if error is retryable
                    let err_str = e.to_string();
                    let retryable = err_str.contains("429")
                        || err_str.contains("502")
                        || err_str.contains("503")
                        || err_str.contains("timeout");

                    if !retryable {
                        return Err(e);
                    }
                    last_error = Some(e);
                    eprintln!(
                        "Request failed (attempt {}/{}), retrying...",
                        attempt + 1,
                        MAX_ATTEMPTS
                    );
                }
            }
        }

        Err(last_error
            .unwrap_or_else(|| anyhow::anyhow!("All {MAX_ATTEMPTS} attempts failed")))
    }

    async fn call_api(&self, text: &str) -> Result<String> {
        match self.provider {
            CloudProvider::OpenAI => self.call_openai(text).await,
            CloudProvider::Gemini => self.call_gemini(text).await,
        }
    }

    async fn call_openai(&self, text: &str) -> Result<String> {
        let url = "https://api.openai.com/v1/chat/completions";
        let schema = serde_json::json!({
            "type": "object",
            "required": ["edits"],
            "properties": {
                "edits": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "required": ["start_index", "end_index", "replacement", "category"],
                        "properties": {
                            "start_index": {"type": "integer", "minimum": 0},
                            "end_index": {"type": "integer", "minimum": 1},
                            "replacement": {"type": "string"},
                            "category": {"type": "string", "enum": ["grammar", "style"]},
                            "explanation": {"type": "string"}
                        },
                        "additionalProperties": false
                    }
                }
            },
            "additionalProperties": false
        });

        let body = json!({
            "model": self.model,
            "messages": [
                {"role": "system", "content": GRAMMAR_CHECK_SYSTEM_PROMPT},
                {"role": "user", "content": format_user_message(text)}
            ],
            "response_format": {
                "type": "json_schema",
                "json_schema": {
                    "name": "grammar_check",
                    "strict": true,
                    "schema": schema
                }
            }
        });

        let response = self
            .client
            .post(url)
            .header("Authorization", format!("Bearer {}", self.api_key))
            .json(&body)
            .send()
            .await
            .context("Failed to send request to OpenAI")?;

        let status = response.status();
        if status == 401 {
            anyhow::bail!("OpenAI authentication failed. Check your API key with `grammar-check config --set-api-key`");
        }
        if !status.is_success() {
            let body = response.text().await.unwrap_or_default();
            anyhow::bail!("OpenAI returned HTTP {status}: {body:.300}");
        }

        let resp: serde_json::Value = response
            .json()
            .await
            .context("Failed to parse OpenAI response")?;

        resp["choices"][0]["message"]["content"]
            .as_str()
            .map(|s| s.to_string())
            .ok_or_else(|| anyhow::anyhow!("Unexpected OpenAI response shape: {resp}"))
    }

    async fn call_gemini(&self, text: &str) -> Result<String> {
        let url = format!(
            "https://generativelanguage.googleapis.com/v1beta/models/{}/generateContent?key={}",
            self.model, self.api_key
        );

        let schema = serde_json::json!({
            "type": "OBJECT",
            "required": ["edits"],
            "properties": {
                "edits": {
                    "type": "ARRAY",
                    "items": {
                        "type": "OBJECT",
                        "required": ["start_index", "end_index", "replacement", "category"],
                        "properties": {
                            "start_index": {"type": "INTEGER"},
                            "end_index": {"type": "INTEGER"},
                            "replacement": {"type": "STRING"},
                            "category": {"type": "STRING", "enum": ["grammar", "style"]},
                            "explanation": {"type": "STRING"}
                        }
                    }
                }
            }
        });

        let body = json!({
            "contents": [{
                "parts": [
                    {"text": GRAMMAR_CHECK_SYSTEM_PROMPT},
                    {"text": format_user_message(text)}
                ]
            }],
            "generationConfig": {
                "responseMimeType": "application/json",
                "responseSchema": schema,
                "temperature": 0.1
            }
        });

        let response = self
            .client
            .post(&url)
            .json(&body)
            .send()
            .await
            .context("Failed to send request to Gemini")?;

        let status = response.status();
        if status == 400 || status == 403 {
            let body = response.text().await.unwrap_or_default();
            anyhow::bail!("Gemini request failed (HTTP {status}): {body:.300}");
        }
        if !status.is_success() {
            let body = response.text().await.unwrap_or_default();
            anyhow::bail!("Gemini returned HTTP {status}: {body:.300}");
        }

        let resp: serde_json::Value = response
            .json()
            .await
            .context("Failed to parse Gemini response")?;

        resp["candidates"][0]["content"]["parts"][0]["text"]
            .as_str()
            .map(|s| s.to_string())
            .ok_or_else(|| anyhow::anyhow!("Unexpected Gemini response shape: {resp}"))
    }
}

impl crate::providers::Provider for CloudLlmProvider {
    async fn check(&self, text: &str) -> Result<Vec<Edit>> {
        let json_response = self.call_with_retry(text).await?;
        validator::parse_and_validate(&json_response, text)
    }

    fn provider_name(&self) -> &str {
        self.provider.as_str()
    }

    fn model_name(&self) -> &str {
        &self.model
    }
}
